eth/vm/computation/base_computation.py

```diff
@@ -0,0 +1,354 @@
+from types import TracebackType
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
+from cached_property import cached_property
+
+from eth_typing import (
+    Address,
+)
+from eth_utils import (
+    get_extended_debug_logger,
+)
+
+from eth._utils.datatypes import (
+    Configurable,
+)
+from eth._utils.numeric import (
+    ceil32,
+)
+from eth.abc import (
+    ComputationAPI,
+    MemoryAPI,
+    StackAPI,
+    GasMeterAPI,
+    OpcodeAPI,
+    CodeStreamAPI,
+    MessageComputationAPI,
+    StateAPI,
+)
+from eth.constants import (
+    GAS_MEMORY,
+    GAS_MEMORY_QUADRATIC_DENOMINATOR,
+)
+from eth.exceptions import (
+    VMError,
+)
+from eth.validation import (
+    validate_is_bytes,
+    validate_uint256,
+)
+from eth.vm.logic.invalid import (
+    InvalidOpcode,
+)
+from eth.vm.memory import (
+    Memory,
+)
+from eth.vm.stack import (
+    Stack,
+)
+
+
+def NO_RESULT(computation: ComputationAPI) -> None:
+    """
+    This is a special method intended for usage as the "no precompile found" result.
+    The type signature is designed to match the other precompiles.
+    """
+    raise Exception("This method is never intended to be executed")
+
+
+def memory_gas_cost(size_in_bytes: int) -> int:
+    size_in_words = ceil32(size_in_bytes) // 32
+    linear_cost = size_in_words * GAS_MEMORY
+    quadratic_cost = size_in_words ** 2 // GAS_MEMORY_QUADRATIC_DENOMINATOR
+
+    total_cost = linear_cost + quadratic_cost
+    return total_cost
+
+
+C = TypeVar("C", bound="ComputationAPI")
+
+
+class BaseComputation(ComputationAPI, Configurable, Generic[C]):
```

Comment:
Do you mean `Configurable` and `ComputationAPI` were swapped in order? I don't think it matters much but I can restore the order. I was going for API (abstract class) -> inherited non-abstract class -> and then the generic.

Reply:
I'm going to merge this since it was mostly approved and this seems like a smaller change. I'd like to be able to start rebasing the EOF PR from these changes. I can address some more computation refactoring in the that PR as well. Happy to move this conversation over there.

Reply:
> Do you mean since Configurable and ComputationAPI were swapped in order? 

Yep, that's what I was looking at. I was more worried about it just being a copy/pasta rather than a breaking change for users. I didn't look into it too much, so if it doesn't matter this seems fine to me!